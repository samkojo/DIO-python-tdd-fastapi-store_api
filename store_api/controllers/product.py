from decimal import Decimal
from typing import Annotated, List, Optional
from fastapi import APIRouter, Body, Depends, HTTPException, Path, Query, status
from pydantic import UUID4
from store_api.core.exceptions import DBInsertException, NotFoundException

from store_api.schemas.product import (
    ProductIn,
    ProductOut,
    ProductUpdate,
    ProductUpdateOut,
)
from store_api.usecases.product import ProductUsecase

router = APIRouter(tags=["products"])


@router.post(path="/", status_code=status.HTTP_201_CREATED)
async def post(
    body: ProductIn = Body(...), usecase: ProductUsecase = Depends()
) -> ProductOut:
    try:
        return await usecase.create(body=body)
    except DBInsertException as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=exc.message)


@router.get(path="/{id}", status_code=status.HTTP_200_OK)
async def get(
    id: UUID4 = Path(alias="id"), usecase: ProductUsecase = Depends()
) -> ProductOut:
    try:
        return await usecase.get(id=id)
    except NotFoundException as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=exc.message)


@router.get(path="/", status_code=status.HTTP_200_OK)
async def query(
    price_grater_than: Annotated[
        Optional[Decimal], Query(description="preço maior que", exemple="5.500")
    ] = None,
    price_less_than: Annotated[
        Optional[Decimal], Query(description="preço menor que", exemple="10.500")
    ] = None,
    usecase: ProductUsecase = Depends(),
) -> List[ProductOut]:
    return await usecase.query(price_grater_than, price_less_than)


@router.patch(path="/{id}", status_code=status.HTTP_200_OK)
async def patch(
    id: UUID4 = Path(alias="id"),
    body: ProductUpdate = Body(...),
    usecase: ProductUsecase = Depends(),
) -> ProductUpdateOut:
    return await usecase.update(id=id, body=body)


@router.delete(path="/{id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete(
    id: UUID4 = Path(alias="id"), usecase: ProductUsecase = Depends()
) -> None:
    try:
        await usecase.delete(id=id)
    except NotFoundException as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=exc.message)
