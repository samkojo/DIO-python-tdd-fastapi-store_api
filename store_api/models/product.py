from store_api.models.base import CreateBaseModel
from store_api.schemas.product import ProductIn


class ProductModel(ProductIn, CreateBaseModel):
    ...
