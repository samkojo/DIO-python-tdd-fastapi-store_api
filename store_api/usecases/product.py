from typing import List
from uuid import UUID
from motor.motor_asyncio import AsyncIOMotorClient, AsyncIOMotorDatabase
import pymongo
from store_api.db.mongo import db_client
from store_api.models.product import ProductModel
from store_api.schemas.product import ProductIn, ProductOut, ProductUpdate, ProductUpdateOut
from store_api.core.exceptions import NotFoundException
from bson import Binary, UuidRepresentation

class ProductUsecase:
    def __init__(self) -> None:
        self.client: AsyncIOMotorClient = db_client.get()
        self.database: AsyncIOMotorDatabase = self.client.get_database()
        self.collection = self.database.get_collection("products")

    async def create(self, body: ProductIn) -> ProductOut:
        product_model = ProductModel(**body.model_dump())
        await self.collection.insert_one(product_model.model_dump())

        return ProductOut(**product_model.model_dump())

    async def get(self, id: UUID) -> ProductOut:
        id_binary = Binary.from_uuid(id, UuidRepresentation.STANDARD)
        result = await self.collection.find_one({"id": id_binary})

        if not result:
            raise NotFoundException(message=f"Product not found with filter: {id}")

        return ProductOut(**result)

    async def query(self) -> List[ProductOut]:
        return [ProductOut(**item) async for item in self.collection.find()]

    async def update(self, id: UUID, body: ProductUpdate) -> ProductUpdateOut:
        id_binary = Binary.from_uuid(id, UuidRepresentation.STANDARD)
        result = await self.collection.find_one_and_update(
            filter={"id": id_binary},
            update={"$set": body.model_dump(exclude_none=True)},
            return_document=pymongo.ReturnDocument.AFTER,
        )

        return ProductUpdateOut(**result)

    async def delete(self, id: UUID) -> bool:
        id_binary = Binary.from_uuid(id, UuidRepresentation.STANDARD)
        product = await self.collection.find_one({"id": id_binary})
        if not product:
            raise NotFoundException(message=f"Product not found with filter: {id}")

        result = await self.collection.delete_one({"id": id_binary})

        return True if result.deleted_count > 0 else False


product_usecase = ProductUsecase()
