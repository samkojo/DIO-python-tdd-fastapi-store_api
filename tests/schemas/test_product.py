from pydantic import ValidationError

import pytest
from store_api.schemas.product import ProductIn
from tests.factories import product_data


def test_schemas_return_success():
    data = product_data()
    product = ProductIn.model_validate(data)

    assert product.name == "Iphone 14 Pro Max"


def test_schemas_return_raise(snapshot):
    data = {"name": "Iphone 14 Pro Max", "quantity": 10, "price": 8.500}

    with pytest.raises(ValidationError) as err:
        ProductIn.model_validate(data)

    assert err.value.errors()[0] == snapshot
