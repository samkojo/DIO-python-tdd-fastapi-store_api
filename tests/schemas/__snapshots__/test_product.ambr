# serializer version: 1
# name: test_schemas_return_raise
  dict({
    'input': dict({
      'name': 'Iphone 14 Pro Max',
      'price': 8.5,
      'quantity': 10,
    }),
    'loc': tuple(
      'status',
    ),
    'msg': 'Field required',
    'type': 'missing',
    'url': 'https://errors.pydantic.dev/2.8/v/missing',
  })
# ---
